Change an instance of `.collect::<Vec<_>>().join("")` to `.collect::<String>()`

yoav-lavi opened the pull request:
Hey,
I've noticed that `compiler/rustc_resolve/src/late/diagnostics.rs` uses `.collect::<Vec<_>>().join("");` over `.collect::<String>();`, the latter being shorter, clearer and more performant:

```rs
    criterion.bench_function("bench 1", |bencher| {
        bencher.iter(|| {
            std::iter::repeat("'a, ")
                .take(10)
                .collect::<Vec<_>>()
                .join("");
        })
    });

    criterion.bench_function("bench 2", |bencher| {
        bencher.iter(|| {
            std::iter::repeat("'a, ").take(10).collect::<String>();
        })
    });
```

```
bench 1                 time:   [88.355 ns 88.526 ns 88.704 ns]                         
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
slope  [88.355 ns 88.704 ns] R^2            [0.9910173 0.9909753]
mean   [88.312 ns 88.623 ns] std. dev.      [640.43 ps 940.42 ps]
median [88.168 ns 88.576 ns] med. abs. dev. [587.33 ps 932.58 ps]

bench 2                 time:   [110.31 ns 110.50 ns 110.77 ns]                         
Found 5 outliers among 100 measurements (5.00%)
  5 (5.00%) high mild
slope  [110.31 ns 110.77 ns] R^2            [0.9907452 0.9903230]
mean   [110.85 ns 111.38 ns] std. dev.      [1.1039 ns 1.5752 ns]
median [110.25 ns 110.91 ns] med. abs. dev. [395.71 ps 1.1579 ns]
```

I'm currently working on a [clippy PR](https://github.com/rust-lang/rust-clippy/pull/8573) to lint this issue but thought to create a manual PR for Deno

Thanks for your consideration!

the8472 commented:
> [Merge branch 'master' of](https://github.com/rust-lang/rust/pull/95216/commits/8932cfce3a1094342dd2849b267113a99331eab0) https://github.com/yoav-lavi/rust

This runs afoul of the [no-merge policy](https://rustc-dev-guide.rust-lang.org/git.html#no-merge-policy), you'll have to rebase.

yoav-lavi commented:
@the8472 Should be fine now

yoav-lavi commented:
@jackh726 you're right, I misread the results in this case. 

All other instances of `.collect::<Vec<_>>().join("")` that I checked were slower.
I wonder why it's behaving differently in this case, will look into it. Closing for now.

Thanks!

yoav-lavi commented:
For instance if you take a similar case that Deno had:

```rs
    criterion.bench_function("1", |bencher| {
        bencher.iter(|| {
            let url = "https://google.com".to_owned();
            let split_specifier = url.as_str().split(':');
            split_specifier.skip(1).collect::<String>();
        })
    });

    criterion.bench_function("2", |bencher| {
        bencher.iter(|| {
            let url = "https://google.com".to_owned();
            let split_specifier = url.as_str().split(':');
            split_specifier.skip(1).collect::<Vec<_>>().join("");
        })
    });
  ```
  
  ```
  1                       time:   [138.23 ns 138.56 ns 138.93 ns]                   
Found 9 outliers among 100 measurements (9.00%)
  4 (4.00%) low mild
  2 (2.00%) high mild
  3 (3.00%) high severe
slope  [138.23 ns 138.93 ns] R^2            [0.9871715 0.9869829]
mean   [138.86 ns 139.47 ns] std. dev.      [1.1495 ns 1.9642 ns]
median [138.92 ns 139.42 ns] med. abs. dev. [698.87 ps 1.1745 ns]

2                       time:   [226.80 ns 227.22 ns 227.57 ns]                   
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
slope  [226.80 ns 227.57 ns] R^2            [0.9915521 0.9916662]
mean   [225.44 ns 226.39 ns] std. dev.      [2.0318 ns 2.8526 ns]
median [225.39 ns 226.53 ns] med. abs. dev. [1.7975 ns 2.8586 ns]
```

yoav-lavi commented:
And my own benchmark:

```rs
fn criterion_benchmark(criterion: &mut Criterion) {
    criterion.bench_function("1", |bencher| {
        bencher.iter(|| {
            vec!["hello", "world"]
                .iter()
                .map(|item| item.to_uppercase())
                .collect::<Vec<String>>()
                .join("")
        })
    });
    criterion.bench_function("2", |bencher| {
        bencher.iter(|| {
            vec!["hello", "world"]
                .iter()
                .map(|item| item.to_uppercase())
                .collect::<String>()
        })
    });
}
```

```
1                       time:   [155.79 ns 155.97 ns 156.16 ns]                   
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe
slope  [155.79 ns 156.16 ns] R^2            [0.9971885 0.9971738]
mean   [155.87 ns 156.58 ns] std. dev.      [672.36 ps 2.9345 ns]
median [156.12 ns 156.26 ns] med. abs. dev. [209.02 ps 650.40 ps]

2                       time:   [118.11 ns 118.20 ns 118.28 ns]                   
Found 8 outliers among 100 measurements (8.00%)
  8 (8.00%) high severe
slope  [118.11 ns 118.28 ns] R^2            [0.9987699 0.9987697]
mean   [118.31 ns 118.87 ns] std. dev.      [794.06 ps 2.0461 ns]
median [117.96 ns 118.43 ns] med. abs. dev. [154.11 ps 692.31 ps]
```
